### PR TITLE
Refactor solid api rendering

### PR DESCRIPTION
**Pull request by romanbruelisauer**
[Jira Task](https://awv-informatik.atlassian.net/browse/CAD-3773)
[buerli PR](https://github.com/awv-informatik/buerli/pull/476)

### CODE REVIEW

File: client/src/components/Main.tsx [hunk 1]
```diff
@@ -131,12 +131,13 @@ const Part: React.FC = () => {
       headlessApi.current = api
       try {
         const p = storeApi.getState().examples.objs[storeApi.getState().activeExample].params
-        productOrSolidId.current = await create(api, p, { onSelect, onResume })
+        const id = await create(api, p, { onSelect, onResume })
+        productOrSolidIds.current = cad instanceof solid ? id as number[] : id as number
```
Comment (STINGNAILS) [inline client/src/components/Main.tsx:135]: Why is this change needed? The return type of await create(...) is the same as productOrSolidIds.current anyway

File: client/src/components/Main.tsx [hunk 2]
```diff
@@ -166,17 +167,21 @@ const Part: React.FC = () => {
       if (headlessApi.current && update && params) {
         set({ loading: true })
         try {
-          productOrSolidId.current = await update(
+          productOrSolidIds.current = await update(
             headlessApi.current,
-            productOrSolidId.current,
+            cad instanceof solid ? productOrSolidIds.current as number : productOrSolidIds.current as number[],
             params,
```
Comment (STINGNAILS) [inline client/src/components/Main.tsx:173]: Same as above; These type casts look redundant. Is it only for the sake of explicity?

File: client/src/components/Main.tsx (hunk 1 above)
Comment (romanbruelisauer) [reply client/src/components/Main.tsx]: You're right. I don't remember why I changed this code. I think I had somehow problems with the types. But it seems fine without the cast as well.
See [55f2cea](https://github.com/awv-informatik/buerli-examples/pull/11/commits/55f2cea068c459a65bacba60cea0c4af3010c575)

File: client/src/components/Main.tsx (hunk 2 above)
Comment (romanbruelisauer) [reply client/src/components/Main.tsx]: See also [55f2cea](https://github.com/awv-informatik/buerli-examples/pull/11/commits/55f2cea068c459a65bacba60cea0c4af3010c575)

File: client/src/models/solid/import-step-2.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { ApiNoHistory, solid } from '@buerli.io/headless'
+import data from '../../resources/solid/AWVLogoCube.stp'
+import { Create, Param } from '../../store'
+
+export const paramsMap: Param[] = [].sort((a, b) => a.index - b.index)
+
```
Comment (STINGNAILS) [inline client/src/models/solid/import-step-2.ts:6]: I think `sort` isn't needed here...

Comment (romanbruelisauer) [reply client/src/models/solid/import-step-2.ts:6]: Of course. This code is in the templates to see where params have to be defined and there is always this sort function appended. In this case it's actually not necessary

File: client/src/models/history/CreatePart.ts [hunk 4]
```diff
@@ -17,18 +19,31 @@ export const create: Create = async (apiType, params) => {
...
+//   if (!api) return
+//   const geoms = await api.createBufferGeometry(productId)
+//   return geoms.map(
+//     geom =>
+//       new THREE.Mesh(
+//         geom,
+//         new THREE.MeshStandardMaterial({ color: new THREE.Color('rgb(52, 89, 87)') }),
+//       ),
+//   )
+// }
```
Comment (dm385) [inline client/src/models/history/CreatePart.ts:32]: Does it make sense to keep this code?

File: client/src/models/history/CreatePart.ts [hunk 5]
```diff
@@ -17,18 +19,31 @@ export const create: Create = async (apiType, params) => {
...
-  )
+  const scene = await api.createScene(productId)
+  scene && colorize(scene)
+  return scene
 }
 
+const colorize = (scene: THREE.Scene) => {
+  const customRed = new Color('rgb(203, 67, 22)')
+  setNodesColor('Part', customRed, scene)
+  setNodesTransparency('Part', 0.5, scene)
```
Comment (dm385) [inline client/src/models/history/CreatePart.ts:44]: I assume "Part" is the name of the group that will be created in createScene, right? Has it always been like that?

Couldn't this be solved in another way? E.g. could createScene return several things instead of leaving the caller to search for hardcoded names?

For example:
```
createScene: (solids: VID[]) => Promise<{scene: Scene, part: Group, solids: Group[]}>
```
Where "part" is of course already a child of "scene" and the "solids" are also already children of "part".

File: client/src/models/solid/Profile.ts [hunk 6]
```diff
@@ -30,19 +30,19 @@ export const create: Create = async (apiType, params) => {
...
-  const scene = await api.createScene(solidId)
+  const scene = await api.createScene(solidIds)
   scene && colorize(scene)
   return scene
 }
 
 const colorize = (scene: THREE.Scene) => {
   const customRed = new Color('rgb(203, 67, 188)')
-  setNodesColor('Solid', customRed, scene)
+  setSolidsColor('Solid0', customRed, scene)
```
Comment (dm385) [inline client/src/models/solid/Profile.ts:45]: The same as above: https://github.com/awv-informatik/buerli-examples/pull/11/files#r1039857321

File: client/src/models/history/CreatePart.ts (hunk 4 above)
Comment (romanbruelisauer) [reply client/src/models/history/CreatePart.ts]: Doesn't make sense. You're right. I removed it. See [64653ef](https://github.com/awv-informatik/buerli-examples/pull/11/commits/64653efebb741eea6b47350fbc24f16b2a8bd2da)

File: client/src/models/history/CreatePart.ts (hunk 5 above)
Comment (romanbruelisauer) [reply client/src/models/history/CreatePart.ts:32]: Why not, could be possible, yes.

Comment (romanbruelisauer) [reply client/src/models/history/CreatePart.ts:32]: So far, I always thought the scene is somehow "userland" and we just give some hints how it could be done.

Comment (dm385) [reply client/src/models/history/CreatePart.ts:32]: Since we create the Scene in the function `createScene`, the creation of it is not userland. Therefore the user has to know exactly what names the objects in the scene have to identify them. If he does not know the name he as no change to get for example the "Part".

Comment (dm385) [reply client/src/models/history/CreatePart.ts:32]: And if he would like to access the solids he has to know about the indexed name as well:
```
mesh.name = 'Solid' + index
```

Comment (romanbruelisauer) [reply client/src/models/history/CreatePart.ts]: createScene() in solid and history api return now all the created THREE objects, see latest commits in [buerli](https://github.com/awv-informatik/buerli/pull/476). The naming for solids and meshes has been changed. The name of the solids correspond to their id. Meshes do only have a name if they're not merged, then the name corresponds to their graphic id.

File: client/src/store.ts [hunk 7]
```diff
@@ -24,19 +24,20 @@ export type Create = (
   api: ApiHistory | ApiNoHistory,
   params?: { lastUpdatedParam: number; values: any[] },
   options?: any,
-) => Promise<number>
+) => Promise<number | number[]>
 export type Update = (
   api: ApiHistory | ApiNoHistory,
-  productId: number,
+  productId: number | number[],
```
Comment (dm385) [inline client/src/store.ts:30]: That change seems to lead to several incompatible API calls

image

Comment (romanbruelisauer) [reply client/src/store.ts:30]: Oh right, forgot to call "yarn build" in buerli-examples. Will fix it.

Comment (romanbruelisauer) [reply client/src/store.ts:30]: I've added a condition to check if product id is a single number. See [53f29de](https://github.com/awv-informatik/buerli-examples/pull/11/commits/53f29de58762e627aeac22d90d5f51fdb18b80e8)

**Review by dm385 — APPROVED** [review]
